packages/cli/src/services/project.service.ee.ts

```diff
@@ -211,28 +224,80 @@ export class ProjectService {
+		});
+		await this.clearCredentialCanUseExternalSecretsCache(projectId);
+	}
+
+	/**
+	 * Adds users to a team project with specified roles.
+	 *
+	 * Throws if you the project is a personal project.
+	 * Throws if the relations contain `project:personalOwner`.
+	 */
+	async addUsersToProject(projectId: string, relations: Relation[]) {
+		const project = await this.getTeamProjectWithRelations(projectId);
+		this.checkRolesLicensed(project, relations);
+
+		if (project.type === 'personal') {
+			throw new ForbiddenError("Can't add users to personal projects.");
+		}
+
+		if (relations.some((r) => r.role === 'project:personalOwner')) {
+			throw new ForbiddenError("Can't add a personalOwner to a team project.");
+		}
+
+		await this.projectRelationRepository.save(
+			relations.map((relation) => ({ projectId, ...relation })),
+		);
+	}
+
+	private async getTeamProjectWithRelations(projectId: string) {
+		const project = await this.projectRepository.findOne({
+			where: { id: projectId, type: 'team' },
 			relations: { projectRelations: true },
 		});
+		ProjectNotFoundError.isDefinedAndNotNull(project, projectId);
+		return project;
+	}
 
-		// Check to see if the instance is licensed to use all roles provided
-		for (const r of relations) {
-			const existing = project.projectRelations.find((pr) => pr.userId === r.userId);
+	/** Check to see if the instance is licensed to use all roles provided */
+	private checkRolesLicensed(project: Project, relations: Relation[]) {
+		for (const { role, userId } of relations) {
+			const existing = project.projectRelations.find((pr) => pr.userId === userId);
 			// We don't throw an error if the user already exists with that role so
 			// existing projects continue working as is.
-			if (existing?.role !== r.role && !this.roleService.isRoleLicensed(r.role)) {
-				throw new UnlicensedProjectRoleError(r.role);
+			if (existing?.role !== role && !this.roleService.isRoleLicensed(role)) {
+				throw new UnlicensedProjectRoleError(role);
 			}
 		}
+	}
 
-		await this.projectRelationRepository.manager.transaction(async (em) => {
-			await this.pruneRelations(em, project);
-			await this.addManyRelations(em, project, relations);
-		});
-		await this.clearCredentialCanUseExternalSecretsCache(projectId);
+	async deleteUserFromProject(projectId: string, userId: string) {
+		const projectExists = await this.projectRepository.existsBy({ id: projectId });
+		if (!projectExists) {
+			throw new ProjectNotFoundError(projectId);
+		}
+
+		// TODO: do we need to prevent project owner from being removed?
+		await this.projectRelationRepository.delete({ projectId, userId });
+	}
+
+	async changeUserRoleInProject(projectId: string, userId: string, role: ProjectRole) {
+		const projectUserExists = await this.projectRelationRepository.existsBy({ projectId, userId });
+		if (!projectUserExists) {
+			throw new ProjectNotFoundError(projectId);
+		}
+
+		// TODO: do we need to block any specific roles here?
```

Comment:
Yeah, this function also should only work with team projects and should disallow the personalOwner role.

TODO: add test for this

Reply:
Test added: https://github.com/n8n-io/n8n/blob/cfdb6ae290e649566d6e895546635205cc0d362b/packages/cli/src/services/__tests__/project.service.ee.test.ts#L215-L219